### ISSUE TITLE
FairPlay key rotation feature

### ISSUE THREAD

**Issue by Jason239** [google/shaka-packager#540]
Could you add the FairPlay key rotation feature if it doesn't supported now?

**Comment by kqyang** [comment]
Related to https://github.com/google/shaka-packager/issues/537#issuecomment-453260987.

@Jason239 Do you have a FairPlay license server set up already? Do you have a sample FairPlay license server url?

**Comment by Jason239** [comment]
Yes, but does it relate to license server url and Key rotation packaging?

**Comment by kqyang** [comment]
@Jason239 Yes, that is main issue for FairPlay key rotation support. A common pattern for the FairPlay license server url is https://<url>?key_id=<key_id>. Does your license server url also have this pattern?

**Comment by Jason239** [comment]
We have in-house format to communicate with the FPS license server.
The server issues a license if the player sends a license request by sending data with our format.
Do you need to know that format?

**Comment by kqyang** [comment]
@Jason239 As far as I know, the message format is defined by Apple as Server Playback Context (SPC) message. The only field in SPC payload the client can control is "Asset ID" which is a third-party ID value that tells the key server which content needs to be decrypted. For your in-house format, are you talking about the format of your "Asset ID"?

Anyway, to support key rotation for FairPlay, the stream needs to be encrypted with different keys. The easiest option to tell the FairPlay server to return with different keys is to encode the key id in HLS key uri. That is why I am asking what is the format of your FairPlay key server.

**Comment by Jason239** [comment]
Yes, you are right.
We set HLS key uri field to our format(content id + period index) as Asset ID.
Then the player can use different keys created with content id and period index in our server.
So I mean, our key server makes different keys by using content id and period index.

And when it comes to KMS server, we are using [this format](https://docs.aws.amazon.com/speke/latest/documentation/live-workflow-methods.html).

**Comment by kqyang** [comment]
> We set HLS key uri field to our format(content id + period index) as Asset ID.

Ok, so for the same content, 'content id' is static and 'period index' will be different for different crypto periods. Do you have a sample URL on how it looks like? It will help us decide how we support the generation of these key uris.

**Comment by Jason239** [comment]
Does it mean the playable content URL ?

**Comment by kqyang** [comment]
No, I mean the HLS key uri, which needs to be generated by packager and inserted into the playlists.

**Comment by Jason239** [comment]
Okay, this is the sample. 'Cid0000000000005a00'
The period index at the back is an 8 bytes hexadecimal.

**Comment by kqyang** [comment]
> Cid0000000000005a00

This is not a FairPlay key uri. It should be something like: 
'skd://fairplay.license.com/streaming/getckc?CrmId=&AccountId=&ContentId=<content_Id>&KeyId=<key_id>'.

**Comment by Jason239** [comment]
We create and issue it directly from our server according to the FPS license format.
If the player sends SPC data containing the key uri to our server, then we issue it directly.

**Comment by Jason239** [comment]
like this,
<pre>
                                                             our fps license
player                                                            server
┌─┐                 request                                 ┌─────────────────┐
│ │-------------------------------------------------------> │check spc(get    │
│ │Add a header that our license server requires(user info).│asset id) and    │
│ │  + spc(in assetId:(skd://Cid0000000000005a00)).         │check user info  │
│ │                                                         │in header.       │
│ │<------------------------------------------------------- │generate license.│
└─┘                 response license(ckc)                   └─────────────────┘
</pre>

**Comment by Jason239** [comment]
And as I mentioned before, the package request API is different from requesting a license.
The picture above is for a license request.

**Comment by kqyang** [comment]
@Jason239 I am not sure if I completely understand design. I don't really care what is sent in the SPC payload. I am interested in the URI field.

So if I understand correctly, the URI is set to "skd://Cid0000000000005a00", i.e. something like this in HLS playlist:

```m3u8
#EXT-X-KEY:METHOD=SAMPLE-AES,URI="skd://Cid0000000000005a00",KEYFORMATVERSIONS="1",KEYFORMAT="com.
apple.streamingkeydelivery"
```

Is that correct? What does 'Cid0000000000005a00' mean? Is it the same for every crypto periods?

**Comment by Jason239** [comment]
Right. It's correct. But it is different for each crypto periods as you said before. 
`Ok, so for the same content, 'content id' is static and 'period index' will be different for different crypto periods.`

**Comment by kqyang** [comment]
I see. Thanks for the clarification.

**Comment by truong-hua** [comment]
@kqyang @joeyparrish may I get update regarding this feature as Fairplay is very popular nowaday.

And I think that we should re-open this https://github.com/shaka-project/shaka-packager/issues/415 too. As you can see, it's not easy to implement Multi DRM with key rotation by using either single Widevine or Playready or Fairplay license server because the license server need to talk to each other to sync the key. So I think it is better if we can support a raw key rotation by listening to an external file change or linux pipe. So that a raw key rotation feature witu proper hls key will enable simple AES encryption (like Clearkey) key rotation. Even it is not as secure as DRM but it enough for many use cases.

**Comment by truong-hua** [comment]
To be more clear, I think we shouldn't stick deeply with any license server but provide a standard common way for any one to integrate their expected license management server. So that we can work with any Widevine, Fairplay... or even Clearkey, Marlin... without the effort to manage each separated license server protocol individually.

In my experiences, file change listening is good enough for any purpose.

**Comment by ronak2121** [comment]
We're also interested in this feature for EVENT streams; is there any update? I'm willing to help with contributing the change back to Shaka. We also have a need to support clear lead for the initial parts of the event and then start DRM protection with periodic key rotations. @truong-hua I'm curious if you'd be willing to help code this with me.

I've learned the primary changes Shaka requires to support this are:

1. An implementation of a key fetcher that reads from files, rather than license servers
2. A more robust algorithm in the raw key encryptor for key rotations (we can probably just read from the file)
3. For HLS, inserting EXT-X-MAP tags along with generating a new init segment (would have to be a separate file each time for EVENT streams) with an EXT-X-DISCONTINUITY/EXT-X-KEY each time the key rotates.
4. For DASH, using multi-period manifests

curious what you think @joeyparrish / @kqyang
